### ISSUE TITLE
Improve ConsolidatorScanPriority comparison performance

### ISSUE THREAD

**Issue by starteleport** [QuantConnect/Lean#8451]
#### Rationale
The `ConsolidatorScanPriority` type implements `IComparable` and is used as the priority in the `SubscriptionManager`'s priority queue. Due to the way comparison is handled for the case when `PriorityQueue` is not supplied with `IComparer` instance explicitly, an intermediate `ObjectComparer` is used for the comparison that eventually delegates comparison to `ConsolidatorScanPriority.CompareTo`.

See the following screenshot (PerfView):
![telegram-cloud-document-2-5370574969757521919](https://github.com/user-attachments/assets/c6954cec-9f55-432e-988c-ca584ef1d65c)

#### Potential Solution
The proposed solution is to introduce a new Comparer and supply it to the PrioorityQueue constructor. This yields 5-10 percent improvement in data points per second for all Lean benchmarks on my machine.

The PR is on the way.

#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

**Comment by starteleport** [comment]
@Martin-Molinero, I wanted to discuss something with you.

Looking at `SubscriptionManager` code, I see the potential for improvement in `ScanPastConsolidators` by replacing `PriorityQueue` with plain `List`. The main question here is whether it's critical that we have the order of consolidator scanning dictated by the `Priority`?

Roughly, I'm thinking about something like this:

```
public void ScanPastConsolidators(DateTime newUtcTime, IAlgorithm algorithm)
{
    if (_consolidatorsToAdd != null)
    {
        lock (_threadSafeCollectionLock)
        {
            _consolidatorsToAdd.DoForEach(x => _consolidatorsSortedByScanTime.Add(x.Item1));
            _consolidatorsToAdd = null;
        }
    }

    for (int i = 0; i < _consolidatorsSortedByScanTime.Count; I++)
    {
        var consolidatorToScan = _consolidatorsSortedByScanTime[I];
        var priority = consolidatorToScan.Priority;

        if (priority.UtcScanTime >= newUtcTime)
        {
            continue;
        }

        if (consolidatorToScan.Disposed)
        {
            // consolidator has been removed
            continue;
        }

        if (priority.UtcScanTime != algorithm.UtcTime)
        {
            // only update the algorithm time once, it's not cheap because of TZ conversions
            algorithm.SetDateTime(priority.UtcScanTime);
        }

        if (consolidatorToScan.UtcScanTime <= priority.UtcScanTime)
        {
            // only scan if we still need to
            consolidatorToScan.Scan();
        }
    }
}
```

This change shows drastic improvement in execution time at the expense of sacrificing the exact scan order. From my perspective, we might be fine with that, but I'm not sure.

What do you think?

**Comment by Martin-Molinero** [comment]
Hey @starteleport! yes sadly we need 100% determinism, original impl was actually a list I think and we had to add priority to make sure it would trigger every time the in the same way, and it needs to scale to thousands of entries (imagine hundreds of symbols each with a few indicators/consolidators), that pushed us to the priority queue

**Comment by starteleport** [comment]
@Martin-Molinero could you please help me grasp the need for taking `UtcScanTime` into account when determining order?

I did some research myself, there were two issues and their corresponding pull requests that contributed to the current implementation:
- https://github.com/QuantConnect/Lean/issues/5457
- https://github.com/QuantConnect/Lean/issues/8017

The second one is about introducing consolidator ID to guarantee ordering between consolidators updating at the same time. The rationale is clear to me.

The first issue was devoted to missing consolidation in no data situations. From the issue and the comments it's not obvious to me why `UtcScanTime` is being used in comparison. I can see that we use `UtcScanTime` inside the condition of the `while` loop to skip those consolidators that we don't need at the current time step. But is it the **only** reason?

Now bear with me, the reason why I keep asking is that the time complexity of `ScanPastConsolidators` is `O(n log n)`, where `n` is the number of consolidators. Reason being, in the worst case we need to go through all elements of the queue (`while`) and do `Enqueue` and `Dequeue` on each step which is `O(log n)` each (https://stackoverflow.com/a/74181115/1845402).

If `UtcScanTime` was only included in sorting to limit the number of elements considered, and nothing more, then we could improve the performance in this spot. 

Assuming it is the case, we could get away with `SortedList` which sorts by consolidator ID only. That way, the list would ever change (`O(n)`) if/when consolidators are added/removed. Other times, we'll just need to do a `O(n)` loop through the list, skipping items that we don't need to scan yet. Given that thousands of consolidators are expected, I think it's worth considering at least.

@jaredbroad, tagging you as well as the reviewer of the related PRs

To sum up, I would really love to learn more about why's of this code, and even if my current proposal is not feasible, I might be able to address it another way. 

Just to give an idea: switching to plain `List` (I haven't tried with `SortedList` yet, but I don't expect much difference) yielded more than 2x improvement in `IndicatorRibbonBenchmark` with 50 consolidators.

**Comment by starteleport** [comment]
@Martin-Molinero @jaredbroad
Hi, it's been a while and I know you guys probably been so busy over the holidays, so I just wanted to gently bump this discussion. 

Another idea came to my mind is to make an opt-in behaviour of having a static consolidator scan order that will be determined by the consolidator dependencies collected from the algorithm author.

I roughly see it this way:
- When creating a consolidator, user may specify a dependency on some other consolidator ("parent" consolidator)
- Given that, we may only guarantee users that updates of "parent" consolidators will be observed in "child" consolidator callback if and only if the dependency was declared
- Otherwise, no guarantees are made

This may help to keep scan order static between time steps and greatly improve backtesting speeds.

I would be really happy to hear your thoughts.

If it's not the right time and/or my idea does not play well with your vision, it's fine.
